lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js

```diff
@@ -20,162 +19,184 @@
+/** @typedef {{tagName: 'SCRIPT', url: string, src: string}} ScriptTag */
+/** @typedef {import('../../driver.js')} Driver */
 
 /* istanbul ignore next */
 function installMediaListener() {
+  // @ts-expect-error - inserted in page to track media changes.
   window.___linkMediaChanges = [];
   Object.defineProperty(HTMLLinkElement.prototype, 'media', {
     set: function(val) {
-      window.___linkMediaChanges.push({
+      /** @type {MediaChange} */
+      const mediaChange = {
         href: this.href,
         media: val,
         msSinceHTMLEnd: Date.now() - window.performance.timing.responseEnd,
         matches: window.matchMedia(val).matches,
-      });
+      };
+      // @ts-expect-error - `___linkMediaChanges` created above.
+      window.___linkMediaChanges.push(mediaChange);
 
       return this.setAttribute('media', val);
     },
   });
 }
 
 /**
- * @return {Promise<{tagName: string, url: string, src: string, href: string, rel: string, media: string, disabled: boolean, mediaChanges: {href: string, media: string, msSinceHTMLEnd: number, matches: boolean}}>}
+ * @return {Promise<Array<LinkTag | ScriptTag>>}
  */
 /* istanbul ignore next */
-function collectTagsThatBlockFirstPaint() {
-  return new Promise((resolve, reject) => {
-    try {
-      const tagList = [...document.querySelectorAll('link, head script[src]')]
-        .filter(tag => {
-          if (tag.tagName === 'SCRIPT') {
-            const scriptTag = /** @type {HTMLScriptElement} */ (tag);
-            return (
-              !scriptTag.hasAttribute('async') &&
-              !scriptTag.hasAttribute('defer') &&
-              !/^data:/.test(scriptTag.src) &&
-              !/^blob:/.test(scriptTag.src) &&
-              scriptTag.getAttribute('type') !== 'module'
-            );
-          } else if (tag.tagName === 'LINK') {
-            // Filter stylesheet/HTML imports that block rendering.
-            // https://www.igvita.com/2012/06/14/debunking-responsive-css-performance-myths/
-            // https://www.w3.org/TR/html-imports/#dfn-import-async-attribute
-            const linkTag = /** @type {HTMLLinkElement} */ (tag);
-            const blockingStylesheet = linkTag.rel === 'stylesheet' &&
-              window.matchMedia(linkTag.media).matches && !linkTag.disabled;
-            const blockingImport = linkTag.rel === 'import' && !linkTag.hasAttribute('async');
-            return blockingStylesheet || blockingImport;
-          }
-
-          return false;
-        })
-        .map(tag => {
-          return {
-            tagName: tag.tagName,
-            url: tag.tagName === 'LINK' ? tag.href : tag.src,
-            src: tag.src,
-            href: tag.href,
-            rel: tag.rel,
-            media: tag.media,
-            disabled: tag.disabled,
-            mediaChanges: window.___linkMediaChanges.filter(item => item.href === tag.href),
```

Comment:
most of these were type errors, only defined on `HTMLScriptElement` _or_ `HTMLScriptElement`, not both. The properties not defined on whichever the current `tag` was ended up as `undefined` and were dropped as properties by the jsonifying when coming back over the protocol, so most of the properties were actually optional, not required like the old return type suggested. It seemed easier just to make it a union type of `LinkTag | ScriptTag` instead.

Reply:
> only defined on `HTMLScriptElement` _or_ `HTMLScriptElement`

`HTMLLinkElement`?

Reply:
> `HTMLLinkElement`?

whoops, yes